File: architecture-skills/design-patterns/examples/facade/facade.cs
// Facade pattern in C#.
//
// A plain class composes the subsystem objects and exposes one method
// per use case. The subsystems stay public for callers that need
// fine-grained control.

using System;

public sealed class Amplifier
{
    public void On() => Console.WriteLine("Amplifier: on");
    public void SetVolume(int level) => Console.WriteLine($"Amplifier: volume {level}");
    public void Off() => Console.WriteLine("Amplifier: off");
}

public sealed class Projector
{
    public void On() => Console.WriteLine("Projector: on");
    public void WideScreenMode() => Console.WriteLine("Projector: 16:9 widescreen mode");
    public void Off() => Console.WriteLine("Projector: off");
}

public sealed class StreamingPlayer
{
    public void On() => Console.WriteLine("StreamingPlayer: on");
    public void Play(string movie) => Console.WriteLine($"StreamingPlayer: playing \"{movie}\"");
    public void Stop() => Console.WriteLine("StreamingPlayer: stopped");
    public void Off() => Console.WriteLine("StreamingPlayer: off");
}

public sealed class HomeTheater
{
    private readonly Amplifier _amp;
    private readonly Projector _projector;
    private readonly StreamingPlayer _player;

    public HomeTheater(Amplifier amp, Projector projector, StreamingPlayer player)
    {
        _amp = amp;
        _projector = projector;
        _player = player;
    }

    public void WatchMovie(string movie)
    {
        Console.WriteLine("= Movie night =");
        _amp.On();
        _amp.SetVolume(5);
        _projector.On();
        _projector.WideScreenMode();
        _player.On();
        _player.Play(movie);
    }

    public void EndMovie()
    {
        Console.WriteLine("= Shutting down =");
        _player.Stop();
        _player.Off();
        _projector.Off();
        _amp.Off();
    }
}

public static class Program
{
    public static void Main()
    {
        var theater = new HomeTheater(new Amplifier(), new Projector(), new StreamingPlayer());
        theater.WatchMovie("Inception");
        theater.EndMovie();
    }
}

File: architecture-skills/design-patterns/examples/facade/facade.java
// Facade pattern in Java.
//
// The textbook home of the pattern: a plain class composing subsystem
// objects and exposing one method per use case. The subsystems remain
// public API for callers that need fine-grained control.

public class facade {

    static final class Amplifier {
        void on() { System.out.println("Amplifier: on"); }
        void setVolume(int level) { System.out.println("Amplifier: volume " + level); }
        void off() { System.out.println("Amplifier: off"); }
    }

    static final class Projector {
        void on() { System.out.println("Projector: on"); }
        void wideScreenMode() { System.out.println("Projector: 16:9 widescreen mode"); }
        void off() { System.out.println("Projector: off"); }
    }

    static final class StreamingPlayer {
        void on() { System.out.println("StreamingPlayer: on"); }
        void play(String movie) { System.out.println("StreamingPlayer: playing \"" + movie + "\""); }
        void stop() { System.out.println("StreamingPlayer: stopped"); }
        void off() { System.out.println("StreamingPlayer: off"); }
    }

    static final class HomeTheater {
        private final Amplifier amp;
        private final Projector projector;
        private final StreamingPlayer player;

        HomeTheater(Amplifier amp, Projector projector, StreamingPlayer player) {
            this.amp = amp;
            this.projector = projector;
            this.player = player;
        }

        void watchMovie(String movie) {
            System.out.println("= Movie night =");
            amp.on();
            amp.setVolume(5);
            projector.on();
            projector.wideScreenMode();
            player.on();
            player.play(movie);
        }

        void endMovie() {
            System.out.println("= Shutting down =");
            player.stop();
            player.off();
            projector.off();
            amp.off();
        }
    }

    public static void main(String[] args) {
        HomeTheater theater = new HomeTheater(new Amplifier(), new Projector(), new StreamingPlayer());
        theater.watchMovie("Inception");
        theater.endMovie();
    }
}
